workflow/controller/operator.go

```diff
@@ -1199,6 +1199,14 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, node *wfv1.NodeStatu
 			node.Outputs = &outputs
 		}
 	}
+	if node.Outputs != nil && node.Outputs.ExitCode == nil {
+		for _, c := range pod.Status.ContainerStatuses {
+			if c.Name == common.MainContainerName && c.State.Terminated != nil {
+				node.Outputs.ExitCode = pointer.StringPtr(fmt.Sprint(c.State.Terminated.ExitCode))
+				updated = true
+			}
+		}
+	}
```

Comment:
MUCH better

Reply:
why have buggy code in the executor where it is hard to test when you can have that ~~buggy~~ code in the controller

Reply:
This code should run before PodGC. I think it should be. we need to test the edge cases.

Reply:
This runs whenever the exit code is known (because the container is terminated) but has not been set yet. 

I'm doubling-down on this refactoring.